h2o-core/src/main/java/water/util/MathUtils.java

```diff
@@ -377,17 +377,24 @@ public static boolean equalsWithinRecSumErr(double actual, double expected, int
     return Math.abs(actual - expected) <= (n-1) * Math.ulp(actual) * absum;
   }
 
-  /** Compare 2 doubles within a tolerance
+  /** 
+   * Compare 2 doubles within a tolerance. True iff the numbers difference within a absolute tolerance or
+   * within an relative tolerance.
+   * 
    *  @param a double
    *  @param b double
-   *  @param abseps - Absolute allowed tolerance
-   *  @param releps - Relative allowed tolerance
+   *  @param absoluteTolerance - Absolute allowed tolerance
+   *  @param relativeTolerance - Relative allowed tolerance
    *  @return true if equal within tolerances  */
-  public static boolean compare(double a, double b, double abseps, double releps) {
-    return
-      Double.compare(a, b) == 0 || // check for equality
-      Math.abs(a-b)/Math.max(a,b) < releps ||  // check for small relative error
-      Math.abs(a - b) <= abseps; // check for small absolute error
+  public static boolean compare(double a, double b, double absoluteTolerance, double relativeTolerance) {
+    assert absoluteTolerance >= 0;
+    assert relativeTolerance >= 0;
+    
+    final boolean equal = Double.compare(a, b) == 0;
+    final double absoluteError = Math.abs(a - b);
+    final double relativeError = Math.abs(absoluteError / Math.max(a, b));
```

Comment:
This has a division by zero problem - please fix.

Reply:
No, it hasn't, IMO. In this case you get ` Double.POSITIVE_INFINITY` as `relativeError`

Reply:
That is correct, you do not get an error but the division by zero problem still exists.

The point is to compare the relative difference using a scale of the numbers provided. There are different ways of defining relative difference but the one we typically use is to take the absolute difference of the 2 number and divide by the maximum _absolute_ value of the 2 numbers. Because the 2 numbers are not the same (exact quality) there will be no division by zero.

Reply:
I guess I see your point now. Is `final double relativeError = Math.abs(absoluteError / Math.max(Math.abs(a), Math.abs(b)));`
 what you had in mind?

Reply:
Yes, that is exactly what I had in mind and why I clicked on "request changes". It addresses more than just division by zero, the method is incorrect for negative numbers as well (case of -1e10 and 0.001). This could even change the results of early stopping for metrics that are not zero-bound.

Reply:
@michalkurka anything else to add?